examples/nmf/script.jl

```diff
@@ -0,0 +1,347 @@
+end
+function dual_obj_weights_grad!(∇, X, K, ε, D, G, ρ1)
+    return ∇ .= Dual.ot_entropic_semidual_grad(X, G, ε, K) - D * E_star_grad(-D' * G / ρ1)
+end
+function dual_obj_dict(X, K, ε, Λ, G, ρ2)
+    return sum(Dual.ot_entropic_semidual(X, G, ε, K)) + ρ2 * sum(E_star(-G * Λ' / ρ2))
+end
+function dual_obj_dict_grad!(∇, X, K, ε, Λ, G, ρ2)
+    return ∇ .= Dual.ot_entropic_semidual_grad(X, G, ε, K) - E_star_grad(-G * Λ' / ρ2) * Λ
+end;
+# 
+# The only remaining part of Wasserstein-NMF to implement is the conversion at optimality from the dual variable $G$ to the primal variables $D, \Lambda$. From the results of Theorems 3 and 4 in Rolet et al., we have for $\Lambda$:
+# ```math
+# \Lambda_i = \operatorname{softmax}((-D^\top G)_i / \rho_1),
+# ```
+# and we have for $D$:
+# ```math
+# D_i = \operatorname{softmax}((-G \Lambda^\top)_i / \rho_2). 
+# ```
+#
+function getprimal_weights(D, G, ρ1)
+    return NNlib.softmax(-D' * G / ρ1; dims=1)
+end
+function getprimal_dict(Λ, G, ρ2)
+    return NNlib.softmax(-G * Λ' / ρ2; dims=1)
+end;
+# 
+# We can now implement functions `solve_weights` and `solve_dict` that solve the respective dual problems for the next iterates of `Λ` and `D`. 
+#
+function solve_weights(X, K, ε, D, ρ1; alg, options)
+    opt = optimize(
+        g -> dual_obj_weights(X, K, ε, D, g, ρ1),
+        (∇, g) -> dual_obj_weights_grad!(∇, X, K, ε, D, g, ρ1),
+        zero.(X),
+        alg,
+        options,
+    )
+    return getprimal_weights(D, Optim.minimizer(opt), ρ1)
+end
+function solve_dict(X, K, ε, Λ, ρ2; alg, options)
+    opt = optimize(
+        g -> dual_obj_dict(X, K, ε, Λ, g, ρ2),
+        (∇, g) -> dual_obj_dict_grad!(∇, X, K, ε, Λ, g, ρ2),
+        zero.(X),
+        alg,
+        options;
+        autodiff=:forward,
+    )
+    return getprimal_dict(Λ, Optim.minimizer(opt), ρ2)
+end;
+
+#
+# ## Example: noisy univariate Gaussians
+# 
+# We set up each observation as a mixture of 3 Gaussians with means sampled from `N(6, σ), N(0, σ), N(-6, σ)` respectively, and mixture weights sampled uniformly from `[0, 1]`. The resulting mixture model is normalised to sum to 1 on the discrete domain `coord`. 
+#
+f(x, μ, σ) = exp.(-(x .- μ) .^ 2)
+coord = range(-12, 12; length=100)
+N = 100
+σ = 1
+X = hcat(
+    [
+        rand() * f(coord, σ * randn() + 6, 1) +
+        rand() * f(coord, σ * randn(), 1) +
+        rand() * f(coord, σ * randn() - 6, 1) for _ in 1:N
+    ]...,
+)
+X = simplex_norm!(X);
+# 
+# We visualise the  observations. 
+#
+plot(coord, X; alpha=0.1, color=:blue, title="Input data", legend=nothing)
+# 
+# We can apply NMF with a squared Frobenius loss using the NMF.jl package. We seek `k = 3` components.  This performs poorly, since the pointwise nature of the loss function cannot handle the translational noise in the data.
+# 
+k = 3
+result = nnmf(X, k; alg=:multmse)
+plot(coord, result.W; title="NMF with Frobenius loss", palette=:Set1_3)
```

Comment:
FYI you can set default settings for Plots with `default(palette=:Set1_3)` at the top only once IIRC.

Reply:
Thanks! Suggestion incorporated